File: src/services/orderAccess.js
const { useSecureCookies } = require("../config/env");

const COOKIE_NAME = "cuore_order_access";
const COOKIE_MAX_AGE_SECONDS = 6 * 60 * 60;

function serializeOrderAccessCookie(tradeNo, accessToken) {
  const parts = [
    `${COOKIE_NAME}=${encodeURIComponent(`${tradeNo}.${accessToken}`)}`,
    "Path=/",
    `Max-Age=${COOKIE_MAX_AGE_SECONDS}`,
    "HttpOnly",
    "SameSite=Lax",
  ];

  if (useSecureCookies) {
    parts.push("Secure");
  }

  return parts.join("; ");
}

function getCookieValue(cookieHeader, cookieName) {
  if (!cookieHeader) {
    return null;
  }

  const match = cookieHeader
    .split(";")
    .map((part) => part.trim())
    .find((part) => part.startsWith(`${cookieName}=`));

  if (!match) {
    return null;
  }

  return match.slice(cookieName.length + 1);
}

function parseOrderAccessCookie(cookieHeader) {
  const rawValue = getCookieValue(cookieHeader, COOKIE_NAME);

  if (!rawValue) {
    return null;
  }

  const decoded = decodeURIComponent(rawValue);
  const [tradeNo, accessToken] = decoded.split(".");

  if (!tradeNo || !accessToken) {
    return null;
  }

  return { tradeNo, accessToken };
}

function hasOrderAccess(req, order) {
  const access = parseOrderAccessCookie(req.headers.cookie);

  return Boolean(
    access &&
    access.tradeNo === order.tradeNo &&
    access.accessToken === order.accessToken,
  );
}

module.exports = {
  COOKIE_NAME,
  hasOrderAccess,
  parseOrderAccessCookie,
  serializeOrderAccessCookie,
};

File: app/services/order_access.py
import urllib.parse
from flask import request, current_app

COOKIE_NAME = "cuore_order_access"
COOKIE_MAX_AGE = 6 * 60 * 60  # 6 hours


def set_order_access_cookie(response, trade_no: str, token: str):
    value = urllib.parse.quote(f"{trade_no}.{token}")
    response.set_cookie(
        COOKIE_NAME,
        value,
        max_age=COOKIE_MAX_AGE,
        httponly=True,
        samesite="Lax",
        secure=current_app.config.get("USE_SECURE_COOKIES", False),
        path="/",
    )


def get_order_access(req) -> tuple | None:
    raw = req.cookies.get(COOKIE_NAME)
    if not raw:
        return None
    decoded = urllib.parse.unquote(raw)
    parts = decoded.split(".", 1)
    if len(parts) != 2 or not parts[0] or not parts[1]:
        return None
    return parts[0], parts[1]  # (trade_no, token)


def has_order_access(req, order) -> bool:
    access = get_order_access(req)
    return bool(
        access
        and access[0] == order.trade_no
        and access[1] == order.access_token
    )
